torchft/process_group_test.py

```diff
@@ -122,6 +137,122 @@ def check_tensors(arg: Any) -> None:  # pyre-ignore[2]
     return works
 
 
+def _test_multi_pg(pg: ProcessGroup, rank: int, tensor: torch.Tensor) -> None:
+    """
+    Helper function to test a set of collective operations in settings with multiple
```

Comment:
IMO I'd split these up into individual collective tests, so that it can be easier isolated when there are issues + users will be able to clearly understand which collectives are supported on which fault tolerant process groups.

Reply:
Yeah I agree with the sentiment! I previously tried something like this in https://github.com/pytorch/torchft/pull/103 with `_test_pg`, but this introduced made the execution time much longer since there were now (# of collectives) times as many process groups that need to be spun up and torn down. That approach also added in more complexities that I wasn't comfortable with.

Ideally I could represent it like this:
```
class MultiPgTest:
  def test_all_gather(self):
    tensor_list = ...
    ...
  
  def test_allreduce(self):
    ...
```

but I'm not sure how I could re-use process groups in this way. Open to any suggestions though!

Reply:
@allenwang28 Couldn't you just create the process groups & subprocesses in a `setupClass` method which would allow them to be re-used across unittests? 

TestDistBackend and MultiProcessTestCase in PT-D might provide some pointers for doing something like this: https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/distributed/distributed_test.py#L566

Reply:
FYI there is a test class called `MultiProcContinuousTest` defined in the same test utils file as MultiProcesTestCase that shares a PG across test instances.  It requires having __main__ defined differently for that test file and isn't compatible with hahving MultiProcesTestCase instnaces inside the same file currently, but it is in use in a number of pt-d tests bc it saves a lot of time

Reply:
I tried adding in `subTest` into `_test_multi_pg` to see if it would delineate the scenarios being tested but unfortunately it doesn't seem to play well with `ThreadPoolExecutor`. `subTest` could be a good approach for `_test_pg` though.